clang/lib/Sema/SemaExpr.cpp

```diff
@@ -227,16 +227,17 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
           (!isUnevaluatedContext() && !isConstantEvaluated());
       bool IsEsimdPrivateGlobal = isSYCLEsimdPrivateGlobal(VD);
       // Non-const statics are not allowed in SYCL except for ESIMD or with the
-      // SYCLGlobalVar attribute.
+      // SYCLGlobalVar or SYCLDetailGlobalVariableAllowed attribute.
       if (IsRuntimeEvaluated && !IsEsimdPrivateGlobal && !IsConst &&
           VD->getStorageClass() == SC_Static &&
-          !VD->hasAttr<SYCLGlobalVarAttr>())
+          (!VD->hasAttr<SYCLGlobalVarAttr>() || !VD->hasAttr<SYCLDetailGlobalVariableAllowedAttr>()))
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelNonConstStaticDataVariable;
       // Non-const globals are not allowed in SYCL except for ESIMD or with the
-      // SYCLGlobalVar attribute.
+      // SYCLGlobalVar or SYCLDetailGlobalVariableAllowed attribute.
       else if (IsRuntimeEvaluated && !IsEsimdPrivateGlobal && !IsConst &&
-               VD->hasGlobalStorage() && !VD->hasAttr<SYCLGlobalVarAttr>())
+               VD->hasGlobalStorage() && (!VD->hasAttr<SYCLGlobalVarAttr>()
```

Comment:
Yeah, changes in this file are in the right direction, I think. I think you might need to extend this check - https://github.com/Fznamznon/llvm/blob/b4b0741d3cf2eeac8632052022eff26d9f750cce/clang/lib/Sema/Sema.cpp#L1685 as well (also you must likely need to define a custom default constructor for your mock device_global class definition).